### ISSUE TITLE
Context.typeof on unknown identifier is Unknown instead of error in display mode.

### ISSUE THREAD

**Issue by back2dos** [HaxeFoundation/haxe#6794]
And using `Context.typeExpr` yields a `TLocal`. Is there any way to opt out of this behavior?

**Comment by nadako** [comment]
dev version should yield `TIdent`, but yeah, I see that a macro would want to handle a proper type error

**Comment by back2dos** [comment]
Well, I can see the general usefulness of this and it's probably a good default even for macros. My problem is not being able to produce the error *at all* in display mode. If there were some "strict" mode (e.g. `Context.typeof(e, true)`) or whatever that'do the job for me.

**Comment by Simn** [comment]
The problem is that at the time you call anything like Context.typeof, it's already too late and the compiler has processed various modules in display mode. If we disabled error suppression at that point, you would likely end up getting lots of errors, none of which you were looking for.

We could expose the display mode configuration to macros, but I'm not sure if I want to go down that route.

**Comment by back2dos** [comment]
I'm not interested in unrelated long gone errors. I like to try/catch `Context.typeof()` to check things about the concrete argument I'm providing, e.g.:

- `Context.typeof([for (x in $e) x])` to know if `e` can be iterated. Right now I'd have to unify with `Iterable` and if that doesn't work look if it's an instance and has a `length` and unifies with `ArrayAccess` and if it's an abstract look if it has a `length` and `@:arrayAccess`. 
- `Context.typeof(macro $i{name})` to know if `name` is defined in scope. Right now I have to check local vars and then the local class while including instance fields only if the local method is an instance field.

Basically I'm reimplementing parts of the typer in the interpreter, which strikes me as rather silly. Instead I'd like to call the type to tell me if something makes sense or not. Right now it always passes but still actually prints the errors, which is why vshaxe diagnostics report all kinds of errors in code that compiles without any issues.

An alternative would be to expose all kinds of things from the typer into macro context but that would bloat the API. Being able to ask the typer even in display mode whether an expression can be correctly typed or not seems much more convenient to me.

**Comment by Simn** [comment]
How do I reproduce this? I'm trying with this:

```haxe
import haxe.macro.Context;

class Main {
	public static function main() {
		#if !macro
		test().|;
		#end
	}

	macro static function test() {
		try {
			trace(Context.defined("display"));
			var c = Context.typeof(macro $i{"a"});
			trace(c);
			return macro 1;
		} catch(e:Dynamic) {
			trace(e);
			return macro "foo";
		}
	}
}
```

But that gives me the output I expected:

```
Haxe print:
source/Main.hx:12: true

Haxe print:
source/Main.hx:17: Unknown identifier : a
```

**Comment by back2dos** [comment]
Yup, that works for me too. I can reproduce is it with ~10 tink libs. I'll try to reduce that ^^

**Comment by Simn** [comment]
Please check again with latest Haxe, there's a good chance that this is related to #7466.

**Comment by back2dos** [comment]
Can't really test because of https://github.com/HaxeFoundation/haxe/issues/7465 ^^

**Comment by Simn** [comment]
I fixed #7465, please check if anything works now.

**Comment by Simn** [comment]
bump

**Comment by back2dos** [comment]
```haxe
class Tst {
  #if !macro
  static function main() 
    test();
  #end
  static macro function test() {
    try haxe.macro.Context.typeof(macro foo)//diagnostics: Unresolved identifier on `foo`
    catch (e:Dynamic) {}
    return macro null;
  }
}
```

**Comment by back2dos** [comment]
Ok, this works in display mode, but not in diagnostics:

```haxe
import haxe.macro.*;

class Test {
  #if !macro
  static function main() {
    trace(test(true / 'false').|);// gives completion for String
  }
  #end
  macro static function test(e:Expr) {
    return
      try {
        Context.typeof(e);
        macro [];
      }
      catch (e:Dynamic) {
        macro 'false';
      }
  }
}
```

In diagnostics, it says "Bool should be Int" and "String should be Int". And if you use autocompletion to make it `test(true / 'false').charAt(0)` then you get an additional "Array<Unknown<0>> has no field charAt" - while still getting proper autocompletion for `String` on that expression.

**Comment by Simn** [comment]
I'm getting a different problem with this example:

![Code_ctQGWmP4YX](https://user-images.githubusercontent.com/634365/82115355-9627c780-9762-11ea-8b7b-363574892dab.png)

The only diagnostics is the `import haxe.macro.*` being reported as unused, which suggests that the compiler never even looks at the macro context.

**Comment by back2dos** [comment]
The error seems to occur only for js and python targets ^^

**Comment by Simn** [comment]
This reminds me of some recent issue, could you check the state here?
